assets/js/omniSelectEngine.js

```diff
@@ -0,0 +1,394 @@
+      destroyOmniSearch();
+    }
+  });
+  // Event for init and destroy
+  $("body").on("keydown", function (e) {
+    const keyCode = e.keyCode;
+    switch (keyCode) {
+      case 27: // escape key.
+        destroyOmniSearch();
+        break;
+      case 190: // dot key.
+        if (!$("input, textarea").is(":focus")) {
+          initOmniSearch();
+        }
+        break;
+    }
+  });
+
+  // Init select2, get data, set events.
+  function initOmniSearch() {
+    $("body").addClass("prevent-scroll");
+    omniSelectElement.addClass("loading");
+    if ($(".omni-search").hasClass("hidden") === false) {
+      return false;
+    }
+
+    $.fn.select2.amd.require(["select2/selection/search"], function (Search) {
+      var oldRemoveChoice = Search.prototype.searchRemoveChoice;
+
+      Search.prototype.searchRemoveChoice = function () {
+        oldRemoveChoice.apply(this, arguments);
+        this.$search.val("");
+      };
+
+      // Init select2
+      omniSelectElement
+        .select2({
+          multiple: true,
+        })
+        .on("select2:select", function (e) {
+          const data = e.params.data;
+          $(".select2-search__field").val(data.text);
+          switch (data.type) {
+            case "project":
+              $(".selected-value").text(data.text);
+              reinitOmniSearchForType("project", data);
+              break;
+            case "task":
+              $(".selected-value").text(data.text);
+              reinitOmniSearchForType("task", data);
+              break;
+            case "taskAction":
+              var ticketId = e.params.data.id;
+              action = e.params.data.action;
+              switch (action) {
+                case "goto":
+                  path = "?tab=ticketdetails#/tickets/showTicket/" + ticketId;
+                  window.location.href = path;
+                  destroyOmniSearch();
+                  break;
+                case "logtime":
+                  path = "?tab=timesheet#/tickets/showTicket/" + ticketId;
+                  window.location.href = path;
+                  destroyOmniSearch();
+                  break;
+              }
+              break;
+            case "projectAction":
+              var projectId = e.params.data.id;
+              action = e.params.data.action;
+              switch (action) {
+                case "goto":
+                  path = "/projects/changeCurrentProject/" + projectId;
+                  window.location.href = path;
+                  destroyOmniSearch();
+                  break;
+                case "createnew":
+                  path =
+                    "/projects/changeCurrentProject/" +
+                    projectId +
+                    "#/tickets/newTicket";
+                  window.location.href = path;
+                  destroyOmniSearch();
+                  break;
+              }
+              break;
+          }
+        });
+      getOmnisearchData();
```

Comment:
This call seams to be places in the middle of every thing?

Reply:
It's an async call to get the data for the select2 after it has been initialized. Moved it further down so it's not all alone up there :/